students/Dustin_L/lesson04/mailroom2.py

```diff
@@ -0,0 +1,219 @@
+#!/usr/bin/env python3
+"""Mail Room 2 Module
+
+This module contains all of the functions for the updated Mail Room 2 module.
+"""
+
+from collections import defaultdict
+import datetime
+import copy
+
+THANK_YOU_OPT = 1
+REPORT_OPT = 2
+LETTERS_OPT = 3
+QUIT_OPT = 4
+
+GIFTS_KEY = 'Gifts'
+NUM_GIFTS_KEY = 'Number of Gifts'
+TOTAL_KEY = 'Total'
+AVE_KEY = 'Average'
+DEFAULT_DICT = {GIFTS_KEY: [],
+                NUM_GIFTS_KEY: 0,
+                TOTAL_KEY: 0,
+                AVE_KEY: 0}
+DONOR_DB = defaultdict(lambda: copy.deepcopy(DEFAULT_DICT),
+                       {'Toni Morrison': {GIFTS_KEY: [1000, 5000, 10000],
+                                          NUM_GIFTS_KEY: 0,
+                                          TOTAL_KEY: 0,
+                                          AVE_KEY: 0},
+                        'Mike McHargue': {GIFTS_KEY: [12000, 50000, 27000],
+                                          NUM_GIFTS_KEY: 0,
+                                          TOTAL_KEY: 0,
+                                          AVE_KEY: 0},
+                        "Flannery O'Connor": {GIFTS_KEY: [38734, 6273, 67520],
+                                              NUM_GIFTS_KEY: 0,
+                                              TOTAL_KEY: 0,
+                                              AVE_KEY: 0},
+                        'Angela Davis': {GIFTS_KEY: [74846, 38470, 7570, 50],
+                                         NUM_GIFTS_KEY: 0,
+                                         TOTAL_KEY: 0,
+                                         AVE_KEY: 0},
+                        'Bell Hooks': {GIFTS_KEY: [634547, 47498, 474729, 4567],
+                                       NUM_GIFTS_KEY: 0,
+                                       TOTAL_KEY: 0,
+                                       AVE_KEY: 0}})
+
+THANK_YOU_FMT = ('\nDear {:s},\n'
+                 'Thank you for your generous donation of ${:.2f}.\n'
+                 '\t\tSincerely,\n'
+                 '\t\t  -Your conscience')
+
+
+def get_usr_input():
+    """Get input from user.
+
+    Prompt user to select one of three choices. If the user selects one of
+    these three, that value is returned. If not, the user is prompted again to
+    select.
+
+    Returns:
+        int: Value corresponding to user choice
+    """
+    select_prompt = ('\nPlease select from the following options:\n'
+                     '\t1. Send a Thank You\n'
+                     '\t2. Create a Report\n'
+                     '\t3. Send letters to all donors\n'
+                     '\t4. quit\n'
+                     ' --> ')
+    usr_in = int(input(select_prompt))
+
+    while usr_in not in (THANK_YOU_OPT, REPORT_OPT, LETTERS_OPT, QUIT_OPT):
```

Comment:
I think we can take it a step further here 😉  let's learn about a cool function `zip`
ref: https://docs.python.org/3.3/library/functions.html#zip

here's a working demonstration:
```
>>> def hello1(): print('hello1')
...
>>> def hello2(): print('hello2')
...
>>> keys = (1, 2)
>>> mapping = dict(zip(keys, (hello1, hello2)))
>>> mapping[1]()
hello1
>>> mapping[2]()
hello2
```

now how can we utilize it in your code?

```
PROMPT_OPTS = (1, 2, 3, 4)
....
while usr_in not in PROMP_OPTS:  # line 70
    print(f"Please try again. Valid options are: {PROMPT_OPTS}")


opt_dict = dict(zip(PROMPT_OPTS, (send_thank_you, create_report, send_letters, quit_mailroom)))  # line 201

....
```